docs/proposals/resource-qos.md

```diff
@@ -0,0 +1,192 @@
+
+Add kubelet/qos/policy.go
+- Decides which memory QoS class a container is in (based on the policy described above)
+- Decides what OOM score all processes in a container should get
+
+Change memory overcommit mode
+- Right now overcommit mode is off on the machines we set up, so if there isn’t enough memory malloc will return null. This prevents QoS, because best-effort containers won’t be killed. Instead, when there isn’t enough memory, and guaranteed containers call malloc, they may not get the memory they want. We want memory guaranteed containers to get the memory they request, and force out memory best-effort containers.
+- Change the memory overcommit mode to 1, so that using excess memory starts the OOM killer. The implication is that malloc won't return null, a process will be killed instead.
+
+Container OOM score configuration
+- We’re focusing on Docker in this implementation (not RKT)
+- OOM scores
+  - Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by root). Processes with higher OOM scores are killed.
+  - The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will always be OOM killed before B.
+  - The final OOM score of a process is also between 0 and 1000
+- Memory best-effort
+  - Set OOM_SCORE_ADJ: 1000
+  - So processes in best-effort containers will have an OOM_SCORE of 1000
+- Memory guaranteed
+  - Set OOM_SCORE_ADJ: -999
+  - So processes in guaranteed containers will have an OOM_SCORE of 0 or 1
+- Memory burstable
+  - If total memory request > 99.8% of available memory, OOM_SCORE_ADJ: 2
+  - Otherwise, set OOM_SCORE_ADJ to 1000 - 10 * (% of memory requested)
+    - This ensures that the OOM_SCORE of burstable containers is > 1
+    - So burstable containers will be killed if they conflict with guaranteed containers
+    - If a burstable container uses less memory than requested, its OOM_SCORE < 1000
+    - So best-effort containers will be killed if they conflict with burstable containers using less than requested memory
+    - If a process in a burstable container uses more memory than the container requested, its OOM_SCORE will be 1000, if not its OOM_SCORE will be < 1000
+    - Assuming that a container typically has a single big process, if a burstable container that uses more memory than requested conflicts with a burstable container using less memory than requested, the former will be killed
+    - If burstable containers with multiple processes conflict, then the formula for OOM scores is a heuristic, it will not ensure "Request and Limit" guarantees. This is one reason why control loops will be added in subsequent iterations.
+- Pod infrastructure container
+  - OOM_SCORE_ADJ: -999
```

Comment:
Personally, I think it should be -1000 until we have better protection for Kubelet (and that's what was in my initial proposal), but @vmarmol suggested that it should be -999. If it's -1000 we run the risk of system panics, for example if the Kubelet uses too much memory.